src/esp32/pwm.rs

```diff
@@ -0,0 +1,209 @@
+    }
+
+    fn take_timer(&mut self, frequency_hz: u32, channel: PwmChannel) -> anyhow::Result<usize> {
+        let timer_number = if self.timer_frequencies.iter().all(|&x| x.is_some()) {
+            // in the future we might want to instead re-purpose the least recently used timer?
+            let err_string = format!(
+                "no more than 4 different PWM frequencies simultaneously allowed, available frequencies: {}, {}, {}, {}",
+                self.timer_frequencies[0].unwrap(),
+                self.timer_frequencies[1].unwrap(),
+                self.timer_frequencies[2].unwrap(),
+                self.timer_frequencies[3].unwrap(),
+            );
+            self.find_timer_by_frequency(frequency_hz).context(err_string)?
+        } else {
+            let timer_number = self
+                .timer_frequencies
+                .iter()
+                .position(|&x| match x {
+                    Some(x) => x == frequency_hz,
+                    None => false
+                });
+            match timer_number {
+                Some(timer_number) => {
+                    self.timer_channels[timer_number].push(channel);
+                    timer_number
+                },
+                None => {
+                    let timer_number = self.timer_channels.iter().position(|x| x.is_empty()).unwrap();
+                    self.timer_frequencies[timer_number] = Some(frequency_hz);
+                    self.timer_channels[timer_number].push(channel);
+                    timer_number
+                }
+            }
+        };
+        Ok(timer_number)
+    }
+
+    fn drop_timer(&mut self, timer_number: usize, channel: PwmChannel) {
+        let timer_channels = &mut self.timer_channels[timer_number];
+        if let Some(index) = timer_channels.iter().position(|&x| x == channel) {
+            timer_channels.swap_remove(index);
+        };
+    }
+
+    fn take_channel(&mut self, n: i32, pin: i32) -> anyhow::Result<PwmChannel> {
+        if let Some(pin_using_channel) = self.channel_pin_subscriptions.get(&n) {
+            anyhow::bail!("CHANNEL{:} already in use by pin {:}", n, pin_using_channel)
+        }
+        let res = match n {
+            0 => PwmChannel::C0,
+            1 => PwmChannel::C1,
+            2 => PwmChannel::C2,
+            3 => PwmChannel::C3,
+            4 => PwmChannel::C4,
+            5 => PwmChannel::C5,
+            6 => PwmChannel::C6,
+            7 => PwmChannel::C7,
+            _ => anyhow::bail!("Could not find CHANNEL{:}", n),
+        };
+        self.channel_pin_subscriptions.insert(n, pin);
+        Ok(res)
+    }
+
+    fn take_next_channel(&mut self, pin: i32) -> anyhow::Result<PwmChannel> {
+        for i in 0..2 {
+            match self.take_channel(i, pin) {
+                Ok(ret) => {
+                    return Ok(ret);
+                }
+                Err(err) => {
+                    log::debug!("Skipping: {:?}", err);
```

Comment:
skipping?

Reply:
I put it in while debugging and left it in by accident